### DISCUSSION TITLE
Modifying the inverted index

### DISCUSSION THREAD

**Discussion by abg0148** [castorini/pyserini#2046] [Q&A]
Hello, 
I was working on a usecase where i wanted to access the inverted index list and modify it to run some experiments. so far i am unable to figure it out. Can someone help out with this usecase. something basic like taking the index and removing certain documents from the postings list, if someone can help share how i can do this. 

Thanks!

**Comment by abg0148** [comment]
i am working on the ms marco dataset

**Reply by lintool** [reply]
Lucene _does_ provide methods for removing documents, but it's not exposed in Anserini, unfortunately...

You can always take the corpus, remove the documents, and then re-index? With MS MARCO v1, indexing time is pretty fast, so for research purposes this approach should work fine...

**Reply by abg0148** [reply]
Wouldnt exactly work for me, since same documents might not get pruned for all inverted lists. I will try using pylucene to update the index and then maybe run evaluations using pyserini.
But this seems like a good feature, what do you say @lintool? I can work on this maybe, if you have some time, maybe we can discuss?

**Reply by lintool** [reply]
Sure, happy to discuss - send a PR?

**Reply by abg0148** [reply]
Will do!

**Comment by GioiaZheng** [comment]
If by "modify the postings list" you mean mutating an existing Lucene index in place, I would not treat that as a supported Pyserini workflow.

`LuceneIndexReader` is useful for inspecting the index: `get_postings_list(...)` lets you traverse postings for a term, and `get_document_vector(...)` lets you inspect a document's term vector. But those APIs are read-side accessors. Pyserini's normal abstraction is: build an index from a collection, then search/read it. It is not designed as a safe editor for Lucene segment internals.

For MS MARCO-style experiments, I would choose the workaround based on what you are trying to ablate:

- If documents should be removed from the retrieval universe, filter the collection first and build a new index.
- If documents should only be ignored for evaluation, keep the index unchanged and filter the run file before evaluation.
- If you want term/posting-level ablations, generate a transformed corpus or transformed per-document text, then reindex that derived collection.

The main reason is consistency. A postings list is not isolated metadata; it is tied to document lengths, collection statistics, term statistics, norms, and segment-level structures used by scoring. Editing only one postings list would make the index internally inconsistent unless all related statistics were rebuilt as well.

So the practical answer is: use `LuceneIndexReader` to inspect postings, but for controlled experiments create a derived corpus/index or post-process rankings, rather than editing postings in the existing index.
